lib/reline/face.rb

```diff
@@ -0,0 +1,124 @@
+class Reline::Face
+  SGR_PARAMETERS = {
+    foreground: {
+      black: 30,
+      red: 31,
+      green: 32,
+      yellow: 33,
+      blue: 34,
+      magenta: 35,
+      cyan: 36,
+      white: 37,
+      bright_black: 90,
+      gray: 90,
+      bright_red: 91,
+      bright_green: 92,
+      bright_yellow: 93,
+      bright_blue: 94,
+      bright_magenta: 95,
+      bright_cyan: 96,
+      bright_white: 97
+    },
+    background: {
+      black: 40,
+      red: 41,
+      green: 42,
+      yellow: 43,
+      blue: 44,
+      magenta: 45,
+      cyan: 46,
+      white: 47,
+      bright_black: 100,
+      gray: 100,
+      bright_red: 101,
+      bright_green: 102,
+      bright_yellow: 103,
+      bright_blue: 104,
+      bright_magenta: 105,
+      bright_cyan: 106,
+      bright_white: 107,
+    },
+    style: {
+      default: 0,
+      bold: 1,
+      faint: 2,
+      italicized: 3,
+      underlined: 4,
+      slowly_blinking: 5,
+      blinking: 5,
+      rapidly_blinking: 6,
+      negative: 7,
+      concealed: 8,
+      crossed_out: 9
+    }
+  }.freeze
+
+  class FaceConfig
+    def initialize(name, &block)
+      @name = name
+      block.call(self)
+      define(:default) unless self.respond_to?(:default)
+    end
+
+    def define(name, foreground: nil, background: nil, style: nil)
+      values = {}
+      values[:foreground] = foreground if foreground
+      values[:background] = background if background
+      values[:style] = style if style
+      sgr = format_to_sgr(values)
+      define_singleton_method(name) { sgr }
```

Comment:
I'm not too sure about the singleton method approach. For example, it means we don't have a way to list all configured variations. I think it'll be easier to debug and maintain if we can use a hash here.

Reply:
>  it means we don't have a way to list all configured variations.

If I don't misunderstand what you say, we do.
`@configs` would be the hash that you say:

```
Reline::Face.instance_variable_get :@configs
=>
{:default=>#<Reline::Face::Config:0x00007fb0a0305400>,
 :completion_dialog=>#<Reline::Face::Config:0x00007fb0a03f6d50>}
```

Reply:
But individual configs have different states/types (e.g. `enhanced`), how can we know what's defined and what are the values?

For example, 

```rb
Reline::Face.config(:my_config) do |face|
  face.define :default, foreground: :blue
  face.define :enhanced, foreground: :red
end
face = Reline::Face[:my_config]
```

I'd like to see something like this when I call `face.inspect`:

```
#<Reline::Face::Config:0x0000000105fc1d30 @states={:default=>{:foreground=>:blue}, :enhanced=>{:foreground=>:red}}>
```

There should also be something like `face.states` to list all registered states.

And to achieve this, the code would look like:


```rb
    def initialize(name, &block)
      @states = {}
      block.call(self)
      define(:default) unless @states[:default]
    end

    def define(name, foreground: nil, background: nil, style: nil)
      values = {}
      values[:foreground] = foreground if foreground
      values[:background] = background if background
      values[:style] = style if style
      sgr = format_to_sgr(values)
      @states[name] = values
    end
```